File: app/src/main/java/com/example/android/evince/adapter/RvMatrixAdapter.java
package com.example.android.evince.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.android.evince.R;
import com.example.android.evince.apputils.AppUtils;
import com.example.android.evince.databinding.ItemMatrixBinding;
import com.example.android.evince.pojo.Matrix;
import com.example.android.evince.utils.Utils;
import com.example.android.evince.viewutils.ViewUtils;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

public class RvMatrixAdapter extends RecyclerView.Adapter {

    private List<Matrix> mList;
    private Context mContext;
    private int color;

    public RvMatrixAdapter(Context mContext, List<Matrix> mList) {
        this.mContext = mContext;
        this.mList = mList;
    }

    public List<Matrix> getList() {
        return mList;
    }

    public void setList(List<Matrix> mList) {
        this.mList = mList;
    }

    public int getColor() {
        return color;
    }

    public void setColor(int color) {
        this.color = color;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_matrix, parent, false);
        ItemMatrixBinding mBinding = ItemMatrixBinding.bind(view);
        return new ItemViewHolder(view, mBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (position != -1) {
            if (holder instanceof ItemViewHolder) {
                ItemViewHolder itemViewHolder = (ItemViewHolder) holder;
                Matrix matrix = mList.get(position);
                itemViewHolder.mBinding.viewMbtn.setText(String.valueOf(matrix.getNumber()));
                ViewUtils.setEnable(matrix.isSelected(), itemViewHolder.mBinding.viewMbtn);
                if (matrix.getColor() != 0 && matrix.getColor() != -1) {
                    itemViewHolder.mBinding.viewMbtn.setBackgroundColor(matrix.getColor());
                } else {
                    itemViewHolder.mBinding.viewMbtn.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.colorWhiteFfA100));
                }
            }
        }
    }

    public Context getContext() {
        return mContext;
    }

    public void setContext(Context mContext) {
        this.mContext = mContext;
    }

    @Override
    public int getItemCount() {
        return Utils.isNotNullNotEmpty(mList) ? mList.size() : 0;
    }

    public void highlightItem(int number, int color) {
        AppUtils.setSelectedInDb(getContext(), number, color);
        List<Matrix> matchedMatrixList = Utils.getFilteredList(mList, input -> input != null && input.getNumber() == number);
        if (Utils.isNotNullNotEmpty(matchedMatrixList)) {
            Matrix matrix = matchedMatrixList.get(0);
            int index = mList.indexOf(matrix);
            if (index != -1) {
                matrix = mList.get(index);
                matrix.setSelected(true);
                matrix.setColor(color);
                updateItem(index, matrix);
            }
        }
    }

    public void updateItem(int position, Matrix matrix) {
        if (Utils.hasElement(mList, position)) {
            mList.set(position, matrix);
            notifyItemChanged(position);
            notifyItemRangeChanged(position, getItemCount());
        }
    }

    public void clearSelection() {
        AppUtils.setSelectedFalseInDb(getContext());
        if (Utils.isNotNullNotEmpty(mList)) {
            for (Matrix matrix : mList) {
                matrix.setSelected(false);
            }
            notifyDataSetChanged();
        }
    }

    public class ItemViewHolder extends RecyclerView.ViewHolder {

        private ItemMatrixBinding mBinding;

        public ItemViewHolder(@NonNull View itemView, ItemMatrixBinding mBinding) {
            super(itemView);
            this.mBinding = mBinding;
            mBinding.viewMbtn.setOnClickListener(view -> {
                int clickedPosition = getAdapterPosition();
                Matrix matrix = mList.get(clickedPosition);
                matrix.setColor(getColor());
                updateItem(clickedPosition, matrix);
            });
        }
    }
}

File: app/src/main/java/com/example/android/evince/adapter/RvMatrixAdapter.kt
package com.example.android.evince.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.example.android.evince.R
import com.example.android.evince.apputils.AppUtils
import com.example.android.evince.databinding.ItemMatrixBinding
import com.example.android.evince.pojo.Matrix
import com.example.android.evince.utils.Utils
import com.example.android.evince.viewutils.ViewUtils
import com.google.common.base.Predicate

class RvMatrixAdapter(var context: Context, private var mList: MutableList<Matrix>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    var color: Int = 0

    /**
     * 11/6/2019
     * Get the current {@see mList} the adapter is currently holding
     *
     * @author srdpatel
     * @since $1.0$
     */
    var list: MutableList<Matrix>
        get() = mList
        set(mList) {
            this.mList = mList
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_matrix, parent, false)
        val mBinding = ItemMatrixBinding.bind(view)
        return ItemViewHolder(view, mBinding)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (position != -1) {
            if (holder is ItemViewHolder) {
                val matrix = mList[position]
                holder.mBinding.viewMbtn.text = matrix.number.toString()
                ViewUtils.setEnable(false, holder.mBinding.viewMbtn)
                if (matrix.color != 0 && matrix.color != -1) {
                    holder.mBinding.viewMbtn.setTextColor(matrix.color)
                    holder.mBinding.viewMbtn.background = ContextCompat.getDrawable(context, R.drawable.dr_white_rect)
                } else {
                    holder.mBinding.viewMbtn.setTextColor(ContextCompat.getColor(context, R.color.colorWhiteFfA100))
                    holder.mBinding.viewMbtn.background = ContextCompat.getDrawable(context, R.drawable.dr_gradient_rect)
                }
            }
        }
    }

    override fun getItemCount(): Int {
        return if (Utils.isNotNullNotEmpty(mList)) mList.size else 0
    }

    /**
     * 11/6/2019
     * Highlights (Sets number and color) for the item
     *
     * @author srdpatel
     * @since $1.0$
     */
    fun highlightItem(number: Int, color: Int) {
        AppUtils.setSelectedInDb(context, number, color)
        val matchedMatrixList = Utils.getFilteredList(mList, Predicate { input -> input != null && input.number == number })
        if (Utils.isNotNullNotEmpty(matchedMatrixList)) {
            var matrix = matchedMatrixList!![0]
            val index = mList.indexOf(matrix)
            if (index != -1) {
                matrix = mList[index]
                matrix.isSelected = true
                matrix.color = color
                updateItem(index, matrix)
            }
        }
    }

    /**
     * 11/6/2019
     * Update the item
     *
     * @author srdpatel
     * @since $1.0$
     */
    fun updateItem(position: Int, matrix: Matrix) {
        if (Utils.hasElement(mList, position)) {
            mList[position] = matrix
            notifyItemChanged(position)
            notifyItemRangeChanged(position, itemCount)
        }
    }

    /**
     * 11/6/2019
     * Re-set each item color to 0 and isSelected field to false
     *
     * @author srdpatel
     * @since $1.0$
     */
    fun clearSelection() {
        AppUtils.setSelectedFalseInDb(context)
        if (Utils.isNotNullNotEmpty(mList)) {
            for (matrix in mList) {
                matrix.color = 0
                matrix.isSelected = false
            }
            notifyDataSetChanged()
        }
    }

    inner class ItemViewHolder(itemView: View, val mBinding: ItemMatrixBinding) : RecyclerView.ViewHolder(itemView)
}
